### PR TITLE
Add Cloth Simulation Filter (CSF)

### LINKED ISSUE #2765: Add Cloth Simulation Filter (CSF) for ground segmentation
Cloth Simulation Filtering (CSF) aims to segregate returns into ground/nonground classes (much like SMRF, PMF, etc.).

> Zhang, Wuming, et al. "An easy-to-use airborne LiDAR data filtering method based on cloth simulation." Remote Sensing 8.6 (2016): 501.

The original code is Apache 2.0 licensed [1]. It has also been integrated as a CloudCompare plugin [2]. It would be quite trivial to also wrap the original implementation as a PDAL filter.

[1] https://github.com/jianboqi/CSF

[2] http://www.cloudcompare.org/doc/wiki/index.php?title=CSF_(plugin)

### PR DESCRIPTION
**Pull request by chambbj**
Close #2765

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -125,6 +125,13 @@ include(${PDAL_CMAKE_DIR}/arbiter.cmake)
 include(${PDAL_CMAKE_DIR}/nlohmann.cmake)
 include(${PDAL_CMAKE_DIR}/openssl.cmake) # Optional
 
+find_package(OpenMP)
+if (OPENMP_FOUND)
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
+endif()
+
```
Comment (hobu) [inline CMakeLists.txt]: Put this in cmake/openmp.cmake

File: filters/private/csf/CSF.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
+// ======================================================================================
```
Comment (hobu) [inline filters/private/csf/CSF.cpp]: The couts in this file should be aliased and set via method so we can wire up PDAL's logging.

**Comment by dbrookes96** [comment]
Cool to see this available in PDAL.

I cobbled together a pretty basic CSF filter myself last year thats more or less the same as this. Mine lacks the ignored values and include last returns options though, which are nice.

There is two things I would like to mention however...

I would recommend adding a try catch around `c.do_filtering` as depending on the bounds of the input points and the cloth resolution CSF will attempt to allocate a **huge** amount of cloth nodes. It doesn't check this internally (maybe it should). Not sure how PDAL traditionally handles these sorts of things though?

In our build of CSF I've removed all uses of `#pragma omp parallel for` as somewhere inside this introduces non-deterministic behavior. An alternative is to set `OMP_NUM_THREADS` to `1`. This was causing our regression tests to fail. CloudCompare has patched one of these away, but it still persists (https://github.com/jianboqi/CSF/issues/21 and https://github.com/CloudCompare/CloudCompare/issues/953). I don't have exact numbers but this didn't seem to completely destroy the performance of CSF. Again, not sure if you guys consider non-deterministic behavior a problem.

Hope this helps.

**Comment by chambbj** [comment]
Thanks for the comments @dbrookes96!

Good news I guess is that we don't actually build with OpenMP support, so I think this non-deterministic behavior concern is somewhat mitigated for the time being. We should consider addressing it more definitively though should someone ever turn OMP back on at some point in the future.

As for the try/catch for the filtering, I can look more closely at how we could guard against potential issues.

**Comment by chambbj** [comment]
@abellgithub @hobu I'm done fiddling with this one for now.

**Comment by hobu** [comment]
:shipit:

**Pull request merged**